File: demo_projects/DataGnome/src/main/kotlin/src/main/kotlin/utils/StringUtils.kt
package utils

object StringUtils {

    fun trim(input: String?): String {
        return input?.trim() ?: ""
    }

    fun split(input: String?, delimiter: String): List<String> {
        return input?.split(delimiter) ?: emptyList()
    }

    fun join(parts: List<String>?, delimiter: String): String {
        return parts?.joinToString(delimiter) ?: ""
    }

    fun toUpperCase(input: String?): String {
        return input?.toUpperCase() ?: ""
    }

    fun toLowerCase(input: String?): String {
        return input?.toLowerCase() ?: ""
    }

    fun capitalizeFirst(input: String?): String {
        return input?.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() } ?: ""
    }

    fun toKebabCase(input: String?): String {
        return input?.split(" ")?.joinToString("-") { it.lowercase() } ?: ""
    }

    fun toSnakeCase(input: String?): String {
        return input?.split(" ")?.joinToString("_") { it.lowercase() } ?: ""
    }

    fun repeat(input: String?, times: Int): String {
        require(times >= 0) { "Repeat count must be non-negative" }
        return input?.repeat(times) ?: ""
    }

    fun contains(input: String?, substring: String): Boolean {
        return input?.contains(substring) ?: false
    }

    fun pad(input: String?, length: Int, padChar: String): String {
        return input?.padStart((input.length + length) / 2, padChar[0])?.padEnd(length, padChar[0]) ?: ""
    }

    fun reverse(input: String?): String {
        return input?.reversed() ?: ""
    }

    fun isPalindrome(input: String?): Boolean {
        val sanitized = input?.filter { it.isLetterOrDigit() }?.lowercase() ?: ""
        return sanitized == sanitized.reversed()
    }

    fun capitalizeWords(input: String?): String {
        return input?.split(" ")?.joinToString(" ") { it.replaceFirstChar { char -> char.uppercase() } } ?: ""
    }

    fun trimStart(input: String?): String {
        return input?.trimStart() ?: ""
    }

    fun trimEnd(input: String?): String {
        return input?.trimEnd() ?: ""
    }

    fun replace(input: String?, oldValue: String, newValue: String): String {
        return input?.replace(oldValue, newValue) ?: ""
    }

    fun substring(input: String?, startIndex: Int, endIndex: Int): String {
        return input?.substring(startIndex, endIndex) ?: ""
    }

    fun indexOf(input: String?, substring: String): Int {
        return input?.indexOf(substring) ?: -1
    }

    fun lastIndexOf(input: String?, substring: String): Int {
        return input?.lastIndexOf(substring) ?: -1
    }

    fun startsWith(input: String?, prefix: String): Boolean {
        return input?.startsWith(prefix) ?: false
    }

    fun endsWith(input: String?, suffix: String): Boolean {
        return input?.endsWith(suffix) ?: false
    }

    fun charAt(input: String?, index: Int): Char {
        return input?.get(index) ?: throw StringIndexOutOfBoundsException("Index out of bounds")
    }

    fun length(input: String?): Int {
        return input?.length ?: 0
    }

    fun format(format: String?, vararg args: Any?): String {
        return if (format != null) String.format(format, *args) else ""
    }

    fun removeWhitespace(input: String?): String {
        return input?.filterNot { it.isWhitespace() } ?: ""
    }

    fun replaceFirst(input: String?, oldValue: String, newValue: String): String {
        return input?.replaceFirst(oldValue, newValue) ?: ""
    }

    fun replaceLast(input: String?, oldValue: String, newValue: String): String {
        return input?.replaceLast(oldValue, newValue) ?: ""
    }

    fun truncate(input: String?, maxLength: Int): String {
        return if (input != null && input.length > maxLength) input.substring(0, maxLength) else input ?: ""
    }

    fun wrapText(input: String?, maxLineLength: Int): String {
        if (input == null) return ""
        val words = input.split(" ")
        val wrappedText = StringBuilder()
        var currentLineLength = 0

        for (word in words) {
            if (currentLineLength + word.length > maxLineLength) {
                wrappedText.append("\n")
                currentLineLength = 0
            }
            if (currentLineLength > 0) {
                wrappedText.append(" ")
                currentLineLength++
            }
            wrappedText.append(word)
            currentLineLength += word.length
        }

        return wrappedText.toString()
    }
}

File: demo_projects/DataGnome/src/main/kotlin/src/utils/StringUtils.js
// StringUtils.js
// Utility class for common string manipulation operations

class StringUtils {
  /**
   * Trims whitespace from both ends of a string.
   * @param {string} str - The string to trim.
   * @returns {string} - The trimmed string.
   */
  static trim(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.trim();
  }
 /**
  * Converts the first character of a string to uppercase.
  * @param {string} str - The string to modify.
  * @returns {string} - The string with the first character in uppercase.
  */
 static capitalizeFirst(str) {
   if (typeof str !== 'string') {
     throw new TypeError('Expected a string');
   }
   return str.charAt(0).toUpperCase() + str.slice(1);
 }
 /**
  * Converts a string to kebab-case.
  * @param {string} str - The string to convert.
  * @returns {string} - The kebab-cased string.
  */
 static toKebabCase(str) {
   if (typeof str !== 'string') {
     throw new TypeError('Expected a string');
   }
   return str
     .replace(/([a-z])([A-Z])/g, '$1-$2')
     .replace(/\s+/g, '-')
     .toLowerCase();
 }
 /**
  * Converts a string to snake_case.
  * @param {string} str - The string to convert.
  * @returns {string} - The snake_cased string.
  */
 static toSnakeCase(str) {
   if (typeof str !== 'string') {
     throw new TypeError('Expected a string');
   }
   return str
     .replace(/([a-z])([A-Z])/g, '$1_$2')
     .replace(/\s+/g, '_')
     .toLowerCase();
 }
 /**
  * Repeats a string a specified number of times.
  * @param {string} str - The string to repeat.
  * @param {number} count - The number of times to repeat the string.
  * @returns {string} - The repeated string.
  */
 static repeat(str, count) {
   if (typeof str !== 'string') {
     throw new TypeError('Expected a string');
   }
   if (typeof count !== 'number' || count < 0) {
     throw new TypeError('Expected a non-negative number');
   }
   return str.repeat(count);
 }
 /**
  * Checks if a string contains a substring.
  * @param {string} str - The string to search within.
  * @param {string} substring - The substring to search for.
  * @returns {boolean} - True if the substring is found, false otherwise.
  */
 static contains(str, substring) {
   if (typeof str !== 'string' || typeof substring !== 'string') {
     throw new TypeError('Expected strings');
   }
   return str.includes(substring);
 }
 /**
  * Pads a string with another string until the resulting string reaches the given length.
  * @param {string} str - The string to pad.
  * @param {number} targetLength - The length of the resulting string once the current string has been padded.
  * @param {string} padString - The string to pad the current string with.
  * @returns {string} - The padded string.
  */
 static pad(str, targetLength, padString = ' ') {
   if (typeof str !== 'string' || typeof padString !== 'string') {
     throw new TypeError('Expected strings');
   }
   if (typeof targetLength !== 'number') {
     throw new TypeError('Expected a number for targetLength');
   }
   return str.padStart((str.length + targetLength) / 2, padString).padEnd(targetLength, padString);
 }

  /**
   * Splits a string into an array of substrings using a specified separator.
   * @param {string} str - The string to split.
   * @param {string} separator - The separator to use for splitting the string.
   * @returns {string[]} - An array of substrings.
   */
  static split(str, separator) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.split(separator);
  }

  /**
   * Converts a string to uppercase.
   * @param {string} str - The string to convert.
   * @returns {string} - The uppercase string.
   */
  static toUpperCase(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.toUpperCase();
  }

  /**
   * Converts a string to lowercase.
   * @param {string} str - The string to convert.
   * @returns {string} - The lowercase string.
   */
  static toLowerCase(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.toLowerCase();
  }
  /**
   * Reverses a string.
   * @param {string} str - The string to reverse.
   * @returns {string} - The reversed string.
   */
  static reverse(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.split('').reverse().join('');
  }
  /**
   * Checks if a string is a palindrome.
   * @param {string} str - The string to check.
   * @returns {boolean} - True if the string is a palindrome, false otherwise.
   */
  static isPalindrome(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    const reversed = str.split('').reverse().join('');
    return str === reversed;
  }
  /**
   * Capitalizes the first letter of each word in a string.
   * @param {string} str - The string to capitalize.
   * @returns {string} - The capitalized string.
   */
  static capitalizeWords(str) {
    if (typeof str !== 'string') {
      throw new TypeError('Expected a string');
    }
    return str.replace(/\b\w/g, char => char.toUpperCase());
  }
}

module.exports = StringUtils;